File: internal/service/sales_service.go
package service

import (
	"fmt"
	"time"

	"github.com/hafizd-kurniawan/pos-baru/internal/domain/models"
	"github.com/hafizd-kurniawan/pos-baru/internal/repository"
)

type SalesService interface {
	CreateTransaction(req *models.SalesTransactionCreateRequest) (*models.SalesTransaction, error)
	GetTransactionByID(id int) (*models.SalesTransaction, error)
	ListTransactions(page, limit int, status, dateFrom, dateTo string, customerID *int) ([]models.SalesTransaction, int64, error)
	UpdateTransaction(id int, req *models.SalesTransactionUpdateRequest) (*models.SalesTransaction, error)
	DeleteTransaction(id int) error
	GetAvailableVehicles(search, brand string, yearFrom, yearTo *int, sortBy, status string) ([]models.Vehicle, error)
}

type salesService struct {
	salesRepo    repository.SalesRepository
	vehicleRepo  repository.VehicleRepository
	customerRepo repository.CustomerRepository
}

func NewSalesService(salesRepo repository.SalesRepository, vehicleRepo repository.VehicleRepository, customerRepo repository.CustomerRepository) SalesService {
	return &salesService{
		salesRepo:    salesRepo,
		vehicleRepo:  vehicleRepo,
		customerRepo: customerRepo,
	}
}

func (s *salesService) CreateTransaction(req *models.SalesTransactionCreateRequest) (*models.SalesTransaction, error) {
	// Validate customer exists
	customer, err := s.customerRepo.GetByID(req.CustomerID)
	if err != nil {
		return nil, fmt.Errorf("customer not found")
	}

	// Validate vehicle exists and available
	vehicle, err := s.vehicleRepo.GetByID(req.VehicleID)
	if err != nil {
		return nil, fmt.Errorf("vehicle not found")
	}

	if vehicle.Status != "available" {
		return nil, fmt.Errorf("vehicle not available for sale")
	}

	// Generate invoice number
	invoiceNumber := s.generateInvoiceNumber()

	// Calculate profit (selling_price - hpp_price)
	profit := req.SellingPrice - vehicle.PurchasePrice

	// Calculate remaining payment
	remainingPayment := req.SellingPrice - req.DownPayment

	// Determine payment status
	paymentStatus := models.PaymentStatusPending
	if req.DownPayment > 0 {
		if remainingPayment > 0 {
			paymentStatus = models.PaymentStatusPartial
		} else {
			paymentStatus = models.PaymentStatusPaid
		}
	}

	if req.PaymentStatus != "" {
		paymentStatus = req.PaymentStatus
	}

	// Create sales transaction
	salesTransaction := &models.SalesTransaction{
		InvoiceNumber:    invoiceNumber,
		TransactionDate:  time.Now(),
		CustomerID:       req.CustomerID,
		VehicleID:        req.VehicleID,
		HPPPrice:         vehicle.PurchasePrice,
		SellingPrice:     req.SellingPrice,
		Profit:           profit,
		PaymentMethod:    req.PaymentMethod,
		PaymentStatus:    paymentStatus,
		DownPayment:      req.DownPayment,
		RemainingPayment: remainingPayment,
		Notes:            req.Notes,
		ProcessedBy:      req.SalespersonID,
	}

	// Save transaction
	savedTransaction, err := s.salesRepo.Create(salesTransaction)
	if err != nil {
		return nil, fmt.Errorf("failed to create sales transaction: %v", err)
	}

	// Update vehicle status to sold
	err = s.vehicleRepo.UpdateStatus(req.VehicleID, "sold")
	if err != nil {
		// Rollback transaction if needed
		return nil, fmt.Errorf("failed to update vehicle status: %v", err)
	}

	// Load relations
	savedTransaction.Customer = customer
	savedTransaction.Vehicle = vehicle

	return savedTransaction, nil
}

func (s *salesService) GetTransactionByID(id int) (*models.SalesTransaction, error) {
	transaction, err := s.salesRepo.GetByID(id)
	if err != nil {
		return nil, fmt.Errorf("sales transaction not found")
	}

	return transaction, nil
}

func (s *salesService) ListTransactions(page, limit int, status, dateFrom, dateTo string, customerID *int) ([]models.SalesTransaction, int64, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 {
		limit = 10
	}

	offset := (page - 1) * limit

	transactions, total, err := s.salesRepo.List(offset, limit, status, dateFrom, dateTo, customerID)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to list sales transactions: %v", err)
	}

	return transactions, total, nil
}

func (s *salesService) UpdateTransaction(id int, req *models.SalesTransactionUpdateRequest) (*models.SalesTransaction, error) {
	// Check if transaction exists
	existingTransaction, err := s.salesRepo.GetByID(id)
	if err != nil {
		return nil, fmt.Errorf("sales transaction not found")
	}

	// Update fields if provided
	if req.SellingPrice != nil {
		existingTransaction.SellingPrice = *req.SellingPrice
		// Recalculate profit
		existingTransaction.Profit = existingTransaction.SellingPrice - existingTransaction.HPPPrice
		// Recalculate remaining payment
		existingTransaction.RemainingPayment = existingTransaction.SellingPrice - existingTransaction.DownPayment
	}

	if req.PaymentMethod != nil {
		existingTransaction.PaymentMethod = req.PaymentMethod
	}

	if req.PaymentStatus != nil {
		existingTransaction.PaymentStatus = *req.PaymentStatus
	}

	if req.DownPayment != nil {
		existingTransaction.DownPayment = *req.DownPayment
		// Recalculate remaining payment
		existingTransaction.RemainingPayment = existingTransaction.SellingPrice - existingTransaction.DownPayment
	}

	if req.Notes != nil {
		existingTransaction.Notes = req.Notes
	}

	// Update timestamp
	existingTransaction.UpdatedAt = time.Now()

	// Save updated transaction
	updatedTransaction, err := s.salesRepo.Update(existingTransaction)
	if err != nil {
		return nil, fmt.Errorf("failed to update sales transaction: %v", err)
	}

	return updatedTransaction, nil
}

func (s *salesService) DeleteTransaction(id int) error {
	// Check if transaction exists
	transaction, err := s.salesRepo.GetByID(id)
	if err != nil {
		return fmt.Errorf("sales transaction not found")
	}

	// Delete transaction
	err = s.salesRepo.Delete(id)
	if err != nil {
		return fmt.Errorf("failed to delete sales transaction: %v", err)
	}

	// Revert vehicle status back to available
	err = s.vehicleRepo.UpdateStatus(transaction.VehicleID, "available")
	if err != nil {
		// Log error but don't fail the delete operation
		fmt.Printf("Warning: failed to revert vehicle status: %v\n", err)
	}

	return nil
}

func (s *salesService) GetAvailableVehicles(search, brand string, yearFrom, yearTo *int, sortBy, status string) ([]models.Vehicle, error) {
	vehicles, err := s.vehicleRepo.SearchAvailable(search, brand, yearFrom, yearTo, sortBy, status)
	if err != nil {
		return nil, fmt.Errorf("failed to get available vehicles: %v", err)
	}

	return vehicles, nil
}

func (s *salesService) generateInvoiceNumber() string {
	now := time.Now()
	return fmt.Sprintf("INV-SALES-%d%02d%02d-%d",
		now.Year(),
		now.Month(),
		now.Day(),
		now.Unix()%10000,
	)
}

File: pos_flutter_app/lib/features/sales/services/sales_service.dart
import '../../../core/models/customer.dart';
import '../../../core/models/vehicle.dart';
import '../../../core/network/api_client.dart';

class SalesTransaction {
  final int? id;
  final String? invoiceNumber;
  final DateTime? transactionDate;
  final int customerId;
  final int vehicleId;
  final double hppPrice;
  final double sellingPrice;
  final double profit;
  final String? paymentMethod;
  final String paymentStatus;
  final double downPayment;
  final double remainingPayment;
  final String? notes;
  final Customer? customer;
  final Vehicle? vehicle;

  SalesTransaction({
    this.id,
    this.invoiceNumber,
    this.transactionDate,
    required this.customerId,
    required this.vehicleId,
    this.hppPrice = 0.0,
    required this.sellingPrice,
    this.profit = 0.0,
    this.paymentMethod,
    required this.paymentStatus,
    required this.downPayment,
    this.remainingPayment = 0.0,
    this.notes,
    this.customer,
    this.vehicle,
  });

  Map<String, dynamic> toJson() {
    return {
      'customer_id': customerId,
      'vehicle_id': vehicleId,
      'hpp_price': hppPrice,
      'selling_price': sellingPrice,
      'profit': profit,
      'payment_method': paymentMethod,
      'payment_status': paymentStatus,
      'down_payment': downPayment,
      'remaining_payment': remainingPayment,
      'notes': notes,
    };
  }

  factory SalesTransaction.fromJson(Map<String, dynamic> json) {
    return SalesTransaction(
      id: json['id'],
      invoiceNumber: json['invoice_number'],
      transactionDate: json['transaction_date'] != null
          ? DateTime.parse(json['transaction_date'])
          : null,
      customerId: json['customer_id'] ?? 0,
      vehicleId: json['vehicle_id'] ?? 0,
      hppPrice: (json['hpp_price'] as num?)?.toDouble() ?? 0.0,
      sellingPrice: (json['selling_price'] as num?)?.toDouble() ?? 0.0,
      profit: (json['profit'] as num?)?.toDouble() ?? 0.0,
      paymentMethod: json['payment_method'],
      paymentStatus: json['payment_status'] ?? 'pending',
      downPayment: (json['down_payment'] as num?)?.toDouble() ?? 0.0,
      remainingPayment: (json['remaining_payment'] as num?)?.toDouble() ?? 0.0,
      notes: json['notes'],
      customer:
          json['customer'] != null ? Customer.fromJson(json['customer']) : null,
      vehicle:
          json['vehicle'] != null ? Vehicle.fromJson(json['vehicle']) : null,
    );
  }
}

class SalesService {
  final ApiClient _apiClient;

  SalesService({required ApiClient apiClient}) : _apiClient = apiClient;

  // Create new sales transaction
  Future<SalesTransaction> createTransaction({
    required int customerId,
    required int vehicleId,
    required double sellingPrice,
    String? paymentMethod,
    String paymentStatus = 'pending',
    double downPayment = 0.0,
    String? notes,
  }) async {
    try {
      final data = {
        'customer_id': customerId,
        'vehicle_id': vehicleId,
        'selling_price': sellingPrice,
        'payment_method': paymentMethod,
        'payment_status': paymentStatus,
        'down_payment': downPayment,
        'notes': notes,
      };

      print('📤 Creating sales transaction: $data');

      final response =
          await _apiClient.post('/api/sales/transactions', data: data);

      if (response.statusCode == 200 || response.statusCode == 201) {
        final responseData = response.data;
        print('✅ Sales transaction created: $responseData');

        if (responseData['data'] != null &&
            responseData['data']['data'] != null) {
          return SalesTransaction.fromJson(responseData['data']['data']);
        }
      }

      throw Exception('Failed to create sales transaction');
    } catch (e) {
      print('❌ Sales transaction creation error: $e');
      throw Exception('Error creating sales transaction: $e');
    }
  }

  // Get available vehicles for sale
  Future<List<Vehicle>> getAvailableVehicles({
    String? search,
    String? brand,
    int? yearFrom,
    int? yearTo,
    String sortBy = 'created_at',
  }) async {
    try {
      String queryParams = '?';
      List<String> params = [];

      if (search != null && search.isNotEmpty) {
        params.add('search=${Uri.encodeComponent(search)}');
      }
      if (brand != null && brand.isNotEmpty) {
        params.add('brand=${Uri.encodeComponent(brand)}');
      }
      if (yearFrom != null) {
        params.add('year_from=$yearFrom');
      }
      if (yearTo != null) {
        params.add('year_to=$yearTo');
      }
      params.add('sort_by=$sortBy');

      queryParams += params.join('&');

      final response =
          await _apiClient.get('/api/sales/vehicles/available$queryParams');

      if (response.statusCode == 200) {
        final data = response.data;
        print('🚗 Available vehicles response: $data');

        if (data['data'] != null) {
          final List<dynamic> vehiclesJson = data['data'];
          return vehiclesJson.map((json) => Vehicle.fromJson(json)).toList();
        }
      }

      return [];
    } catch (e) {
      print('❌ Get available vehicles error: $e');
      return [];
    }
  }

  // Get sales transactions
  Future<List<SalesTransaction>> getTransactions({
    int page = 1,
    int limit = 10,
    String? status,
    String? dateFrom,
    String? dateTo,
    int? customerId,
  }) async {
    try {
      String queryParams = '?page=$page&limit=$limit';

      if (status != null && status.isNotEmpty) {
        queryParams += '&status=${Uri.encodeComponent(status)}';
      }
      if (dateFrom != null && dateFrom.isNotEmpty) {
        queryParams += '&date_from=${Uri.encodeComponent(dateFrom)}';
      }
      if (dateTo != null && dateTo.isNotEmpty) {
        queryParams += '&date_to=${Uri.encodeComponent(dateTo)}';
      }
      if (customerId != null) {
        queryParams += '&customer_id=$customerId';
      }

      final response =
          await _apiClient.get('/api/sales/transactions$queryParams');

      if (response.statusCode == 200) {
        final data = response.data;
        print('📋 Sales transactions response: $data');

        if (data['data'] != null) {
          final List<dynamic> transactionsJson = data['data'];
          return transactionsJson
              .map((json) => SalesTransaction.fromJson(json))
              .toList();
        }
      }

      return [];
    } catch (e) {
      print('❌ Get sales transactions error: $e');
      return [];
    }
  }

  // Get single sales transaction
  Future<SalesTransaction?> getTransaction(int id) async {
    try {
      final response = await _apiClient.get('/api/sales/transactions/$id');

      if (response.statusCode == 200) {
        final data = response.data;
        print('📄 Sales transaction response: $data');

        if (data['data'] != null) {
          return SalesTransaction.fromJson(data['data']);
        }
      }

      return null;
    } catch (e) {
      print('❌ Get sales transaction error: $e');
      return null;
    }
  }
}
